vochain/account.go

```diff
@@ -692,6 +698,97 @@ func (v *State) SetAccountDelegate(accountAddr, delegateAddr common.Address, txT
 	}
 }
 
+// CollectFaucetTxCheck checks if a CollectFaucetTx and its data are valid
+func CollectFaucetTxCheck(vtx *models.Tx, txBytes, signature []byte, state *State) (*common.Address, error) {
+	if vtx == nil {
+		return nil, ErrNilTx
+	}
+	tx := vtx.GetCollectFaucet()
+	// check signature available
+	if signature == nil || tx == nil || txBytes == nil {
+		return nil, fmt.Errorf("missing signature and/or transaction")
+	}
+	faucetPkg := tx.GetFaucetPackage()
+	// check faucet pkg content
+	if faucetPkg == nil {
+		return nil, fmt.Errorf("nil faucet package")
+	}
+	if faucetPkg.Signature == nil {
+		return nil, fmt.Errorf("invalid faucet package signature")
+	}
+	if faucetPkg.Payload == nil {
+		return nil, fmt.Errorf("invalid faucet package payload")
+	}
+	if faucetPkg.Payload.Amount == 0 {
+		return nil, fmt.Errorf("invalid faucet package payload amount")
+	}
+	// recover txSender address from signature
+	pubKey, err := ethereum.PubKeyFromSignature(txBytes, signature)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract public key from signature: %w", err)
+	}
+	toAddr, err := ethereum.AddrFromPublicKey(pubKey)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract address from public key: %w", err)
+	}
+	// extract the account that generated the payload
+	faucetPackageBytes, err := proto.Marshal(faucetPkg.Payload)
+	if err != nil {
+		return nil, fmt.Errorf("cannot extract faucet package payload: %v", err)
+	}
+	fromAddr, err := ethereum.AddrFromSignature(faucetPackageBytes, tx.FaucetPackage.Signature)
+	if err != nil {
+		return nil, err
+	}
+	// check issuer nonce not used
+	b := make([]byte, 8)
+	binary.LittleEndian.PutUint64(b, faucetPkg.Payload.Identifier)
+	key := fromAddr.Bytes()
+	key = append(key, b...)
+	used, err := state.FaucetNonce(crypto.Sha256(key), false)
+	if err != nil {
+		return nil, fmt.Errorf("cannot check faucet nonce: %v", err)
+	}
+	if used {
+		return nil, fmt.Errorf("nonce %d already used", faucetPkg.Payload.Identifier)
+	}
+	// check issuer have enough funds
+	issuerAcc, err := state.GetAccount(fromAddr, false)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get faucet account: %v", err)
+	}
+	if issuerAcc == nil {
+		return nil, fmt.Errorf("the account signing the faucet payload does not exist")
+	}
+	cost, err := state.TxCost(models.TxType_COLLECT_FAUCET, false)
+	if err != nil {
+		return nil, fmt.Errorf("cannot get %s tx cost: %w", models.TxType_COLLECT_FAUCET, err)
+	}
+	if (issuerAcc.Balance) < faucetPkg.Payload.Amount+cost {
+		return nil, fmt.Errorf("faucet does not have enough balance %d < %d", issuerAcc.Balance, faucetPkg.Payload.Amount+cost)
+	}
+	// check tx sender is the same as the one contained in the payload
+	if !bytes.Equal(toAddr.Bytes(), faucetPkg.Payload.To) {
+		return nil, fmt.Errorf("txSender %s and faucet payload To %s missmatch",
```

Comment:
is it safe to use `%s` on these bytes? presumably they may contain non-utf8, so perhaps use %x for hex or %q for quoted Go strings, which will escape non-printable chars

Reply:
typo: s/missmatch/mismatch/